test/integration/oauth_cert_fallback_test.go

```diff
@@ -0,0 +1,189 @@
+// +build integration
+
+package integration
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"testing"
+
+	oclient "github.com/openshift/origin/pkg/client"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/client/restclient"
+
+	"github.com/openshift/origin/pkg/cmd/server/admin"
+	"github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+	testutil "github.com/openshift/origin/test/util"
+	testserver "github.com/openshift/origin/test/util/server"
+)
+
+func TestOAuthCertFallback(t *testing.T) {
+
+	var (
+		invalidToken = "invalid"
+		noToken      = ""
+
+		invalidCert = restclient.TLSClientConfig{
+		// We have to generate this dynamically in order to have an invalid cert signed by a signer with the same name as the valid CA
+		// CertData: ...,
+		// KeyData:  ...,
+		}
+		noCert = restclient.TLSClientConfig{}
+
+		tokenUser = "user"
+		certUser  = "system:admin"
+
+		unauthorizedError = "the server has asked for the client to provide credentials (get users ~)"
+		anonymousError    = `User "system:anonymous" cannot get users at the cluster scope`
+	)
+
+	testutil.RequireEtcd(t)
+	// Build master config
+	masterOptions, err := testserver.DefaultMasterOptions()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Start server
+	clusterAdminKubeConfig, err := testserver.StartConfiguredMaster(masterOptions)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	adminConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	validCert := adminConfig.TLSClientConfig
+
+	validToken, err := tokencmd.RequestToken(adminConfig, nil, tokenUser, "pass")
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	if len(validToken) == 0 {
+		t.Fatalf("Expected valid token, got none")
+	}
+
+	// make a client cert signed by a fake CA with the same name as the real CA.
+	// this is needed to get the go client to actually send the cert to the server,
+	// since the server advertises the signer name it requires
+	fakecadir, err := ioutil.TempDir("", "fakeca")
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	defer os.RemoveAll(fakecadir)
+	cacerts, err := util.CertificatesFromFile(masterOptions.ServingInfo.ClientCA)
+	if err != nil || len(cacerts) != 1 {
+		t.Fatalf("Unexpected error: %v", err)
```

Comment:
Technically this could give an incorrect error message when `err == nil` but `len(cacerts) != 1`.